include/kernels/dslash_domain_wall_4d.cuh

```diff
@@ -8,14 +8,15 @@ namespace quda
   template <typename Float, int nColor, int nDim, QudaReconstructType reconstruct_>
   struct DomainWall4DArg : WilsonArg<Float, nColor, nDim, reconstruct_> {
     typedef typename mapper<Float>::type real;
-    int Ls;                             /** fifth dimension length */
+    int_fastdiv Ls;                             /** fifth dimension length */
     complex<real> a_5[QUDA_MAX_DWF_LS]; /** xpay scale factor for each 4-d subvolume */
 
-    DomainWall4DArg(ColorSpinorField &out, const ColorSpinorField &in, const GaugeField &U, double a, double m_5,
-                    const Complex *b_5, const Complex *c_5, bool xpay, const ColorSpinorField &x, int parity,
+    DomainWall4DArg(cvector_ref<ColorSpinorField> &out, cvector_ref<const ColorSpinorField> &in,
+                    const ColorSpinorField &halo, const GaugeField &U, double a, double m_5,
+                    const Complex *b_5, const Complex *c_5, bool xpay, cvector_ref<const ColorSpinorField> &x, int parity,
                     bool dagger, const int *comm_override) :
-      WilsonArg<Float, nColor, nDim, reconstruct_>(out, in, U, xpay ? a : 0.0, x, parity, dagger, comm_override),
-      Ls(in.X(4))
+      WilsonArg<Float, nColor, nDim, reconstruct_>(out, in, halo, U, xpay ? a : 0.0, x, parity, dagger, comm_override),
+      Ls(in[0].X(4))
```

Comment:
Similar to comments I've made elsewhere, should we make `.X()` a specialized function in the `vector_ref` class definition?

Reply:
3c9dd1609